dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,335 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "common.h"
+#include "Vector.h"
+#include "PipsControl.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include "PipsControlTemplateSettings.h"
+#include "PipsControlSelectedIndexChangedEventArgs.h"
+#include "PipsControlAutomationPeer.h"
+
+
+constexpr auto c_previousPageButtonVisibleVisualState = L"PreviousPageButtonVisible"sv;
+constexpr auto c_previousPageButtonHiddenVisualState = L"PreviousPageButtonHidden"sv;
+
+constexpr auto c_previousPageButtonEnabledVisualState = L"PreviousPageButtonEnabled"sv;
+constexpr auto c_previousPageButtonDisabledVisualState = L"PreviousPageButtonDisabled"sv;
+
+constexpr auto c_nextPageButtonVisibleVisualState = L"NextPageButtonVisible"sv;
+constexpr auto c_nextPageButtonHiddenVisualState = L"NextPageButtonHidden"sv;
+
+constexpr auto c_nextPageButtonEnabledVisualState = L"NextPageButtonEnabled"sv;
+constexpr auto c_nextPageButtonDisabledVisualState = L"NextPageButtonDisabled"sv;
+
+constexpr auto c_infiniteItemsModeState = L"InfiniteItems"sv;
+
+constexpr auto c_rootGridName = L"RootGrid"sv;
+
+constexpr auto c_previousPageButtonName = L"PreviousPageButton"sv;
+constexpr auto c_nextPageButtonName = L"NextPageButton"sv;
+
+constexpr auto c_PipsControlPipsRepeaterName = L"VerticalPipsItemsRepeater"sv;
+constexpr auto c_PipsControlScrollViewerName = L"VerticalPipsScrollViewer"sv;
+
+constexpr auto c_PipsControlButtonHeightPropertyName = L"PipsControlVerticalPipButtonHeight"sv;
+
+PipsControl::PipsControl()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_PipsControl);
+
+    m_verticalPipsElements = winrt::make<Vector<IInspectable>>().as<winrt::IObservableVector<IInspectable>>();
+    const auto templateSettings = winrt::make<PipsControlTemplateSettings>();
+    templateSettings.SetValue(PipsControlTemplateSettings::s_VerticalPipsItemsProperty, m_verticalPipsElements);
+    SetValue(s_TemplateSettingsProperty, templateSettings);
+
+    SetDefaultStyleKey(this);
+
+}
+
+PipsControl::~PipsControl() {
+    m_nextPageButtonClickRevoker.revoke();
+    m_previousPageButtonClickRevoker.revoke();
+    m_verticalPipsElementPreparedRevoker.revoke();
+    m_rootGridPointerEnteredRevoker.revoke();
+    m_rootGridPointerExitedRevoker.revoke();
+    m_rootGridKeyDownRevoker.revoke();
```

Comment:
the revokers should auto revoke, so I don't think we need this destructor

Reply:
Thats correct, these aren't needed

---
In reply to: [520921816](https://github.com/microsoft/microsoft-ui-xaml/pull/3592#discussion_r520921816) [](ancestors = 520921816)